keepcurrent.go

```diff
@@ -40,67 +44,98 @@ type Runner struct {
 
 // New construct a runner which synchronizes data from one source to one or more sinks
 func New(from Source, to ...Sink) *Runner {
-	return &Runner{func(error) {}, func(Sink, error) {}, from, to, time.Time{}}
+	return &Runner{
+		func(error, int) time.Duration { return 0 },
+		func(Sink, error) {},
+		from, to, time.Time{}}
 }
 
 // InitFrom synchronizes data from the given source to configured sinks.
 func (runner *Runner) InitFrom(s Source) {
 	if len(runner.sinks) == 0 {
 		return
 	}
-	runner.syncOnce(s)
+	runner.syncOnce(s, nil)
 }
 
-// Start starts a loop to actually synchronizes data with given interval.  It
+// Start starts the loop to actually synchronizes data with given interval. It
 // returns a function to stop the loop.
 func (runner *Runner) Start(interval time.Duration) func() {
 	if len(runner.sinks) == 0 {
 		return func() {}
 	}
 	tk := time.NewTicker(interval)
-	ch := make(chan struct{})
+	chStop := make(chan struct{})
 	go func() {
 		for {
-			runner.syncOnce(runner.source)
+			runner.syncOnce(runner.source, chStop)
 			select {
-			case <-ch:
+			case <-chStop:
 				tk.Stop()
 				return
 			case <-tk.C:
 			}
 		}
 	}()
-	return func() { close(ch) }
+	return func() { close(chStop) }
```

Comment:
This stop callback can return while `runner.syncOnce` continues to run.  Personally, I would be confused if this were to happen.  [Here's a PR](https://github.com/getlantern/keepcurrent/pull/2) that waits for `syncOnce` to return first.  It assumes good behavior of the source and sink though (they should stop eventually).  An alternative solution might be to check on `chStop` in `syncOnce` even after errors.  Currently, the channel is only checked if there was no error.

Reply:
Good catch. I'll just keep your fix.

Reply:
> Currently, the channel is only checked if there was no error.

The channel is checked when there was error and waited to retry, see https://github.com/getlantern/keepcurrent/pull/1/files#diff-f6f4ad39e1620037805a0b1c38a1e5eaR104

Reply:
Hmm, I misread that somehow.  Sounds good then!